widget_selectplaylist.go

```diff
@@ -0,0 +1,249 @@
+	// m.inputField.SetAutocompletedFunc(func(text string, index int, source int) bool {
+	// 	m.inputField.SetText(text)
+	// 	for _, p := range ui.playlists {
+	// 		if p.Name == text {
+	// 			m.overwrite.SetDisabled(false)
+	// 			m.overwriteEnabled = true
+	// 			m.focusNext(nil)
+	// 			return false
+	// 		}
+	// 	}
+	// 	m.overwrite.SetDisabled(true)
+	// 	m.overwriteEnabled = false
+	// 	return false
+	// })
+	acceptFunc := func() {
+		inputText := m.inputField.GetText()
+		if !m.overwrite.IsChecked() {
+			for _, p := range ui.playlists {
+				if p.Name == inputText {
+					return
+				}
+			}
+		}
+		ui.queuePage.saveQueue(inputText)
+		ui.CloseSelectPlaylist()
+	}
+	m.accept.SetSelectedFunc(acceptFunc)
+	cancelFunc := func() {
+		m.inputField.SetText("")
+		m.overwrite.SetDisabled(true)
+		m.overwriteEnabled = false
+		m.overwrite.SetChecked(false)
+		m.accept.SetDisabled(!m.overwrite.IsChecked())
+		ui.CloseSelectPlaylist()
+	}
+	m.cancel.SetSelectedFunc(cancelFunc)
+
+	buttons := tview.NewFlex().
+		SetDirection(tview.FlexColumn).
+		AddItem(tview.NewFlex(), 0, 1, false).
+		AddItem(m.accept, 0, 4, false).
+		AddItem(tview.NewFlex(), 0, 1, false).
+		AddItem(m.cancel, 0, 4, false).
+		AddItem(tview.NewFlex(), 0, 1, false)
+
+	m.Root = tview.NewFlex().
+		SetDirection(tview.FlexRow).
+		AddItem(m.inputField, 1, 1, true).
+		AddItem(m.overwrite, 1, 1, false).
+		AddItem(buttons, 0, 1, false)
+
+	m.Root.SetInputCapture(func(event *tcell.EventKey) *tcell.EventKey {
+		if m.ui.app.GetFocus() == m.inputField {
+			switch event.Key() {
+			case tcell.KeyTab:
+				if event.Modifiers()&tcell.ModShift != 0 {
+					return m.focusPrev(event)
+				} else {
+					return m.focusNext(event)
+				}
+			case tcell.KeyBacktab:
+				return m.focusPrev(event)
+			case tcell.KeyESC:
+				cancelFunc()
+			}
+			return event
+		}
+		if event.Rune() == ' ' {
+			focused := m.ui.app.GetFocus()
+			if focused == m.accept {
+				acceptFunc()
+				return nil
+			}
+			if focused == m.cancel {
+				cancelFunc()
+				return nil
+			}
+			return event
+		}
+		switch event.Key() {
+		case tcell.KeyESC:
+			cancelFunc()
+			return nil
+		case tcell.KeyCR:
+			focused := m.ui.app.GetFocus()
+			if focused == m.accept {
+				acceptFunc()
+				return nil
+			} else if focused == m.cancel {
+				cancelFunc()
+				return nil
+			}
+			m.focusNext(event)
+			return event
+		case tcell.KeyTab:
+			if event.Modifiers()&tcell.ModShift != 0 {
+				return m.focusPrev(event)
+			} else {
+				return m.focusNext(event)
+			}
+		case tcell.KeyBacktab:
+			return m.focusPrev(event)
+		case tcell.KeyDown:
+			return m.focusNext(event)
+		case tcell.KeyUp:
+			return m.focusPrev(event)
+		default:
+			m.ui.logger.Printf("non-input key = %d", event.Key())
+		}
+		return event
+	})
+
+	m.Root.Box.SetBorder(true).SetTitle(" Playlist Name ")
+
+	return
+}
+
+func (m *PlaylistSelectionWidget) focusNext(event *tcell.EventKey) *tcell.EventKey {
```

Comment:
Is there a way to make this implementation of a focus hierarchy less complex?
With a cursory search I couldn't find any info about tcell/tview being able to handle this for us. Weird.

Reply:
I wish! This dialog made the patch so much harder. I think `tview` simply doesn't provide high-level constructs, like dialogs; much as `tview` itself provides more complex UI elements on top of `tcell`, I think something that provides more complex widgets such as wizards and dialogs is needed. Bubbletea does this sort of thing, stacking different libraries to provide quite sophisticated UI elements.

If I'd had a clear use case for another need for a dialog widget, I'd have factored it out into a more abstract, re-usable structure. This would have necessitated more code and more complexity, and without a clear need it felt like eager optimization. I think if we come across another need for a dialog like this, then it definitely should be factored out into a seperate, isolated widget. Or, maybe there's a library built on `tview` that already provides dialog functionality; TBH I didn't think to go look for it -- when I started, like you I _assumed_ `tview` would provide this, and it took me a while to discover that I had to implement all this myself.

I agree with you: it's complex, and extremely tedious, and feels fragile. It certainly took forever to test and make sure all the edge cases were covered.

TL;DR I'm all for factoring it out. However, I think (a) we should see if someone's already written some higher-level widgets on `tview` and use that, and (b) wait until the next feature that needs a dialog. How does that sound?

Reply:
> Is there a way to make this implementation of a focus hierarchy less complex? With a cursory search I couldn't find any info about tcell/tview being able to handle this for us. Weird.

Agreed.

Reply:
https://github.com/navidys/tvxwidgets, and in particular, https://github.com/navidys/tvxwidgets/blob/main/dialog.go

Also, most recently committed to 13 hours ago.  151 projects using it, 10 contributors. I've not seen this before; I may try re-implementing this patch with that library. 7 new dependencies, and it uses a different log library than we do, which is always a PITA.

Reply:
> https://github.com/navidys/tvxwidgets, and in particular, https://github.com/navidys/tvxwidgets/blob/main/dialog.go

That library looks insanely cool though. We could have a real progress bar and volume sliders!

Reply:
> That library looks insanely cool though. We could have a real progress bar and volume sliders!

I know, right?? I'm already thinking of how much bespoke code we could factor out.

This, and the opensubsonic library -- I think we should look at these first thing post-1.0.0.